### ISSUE TITLE
How do you actually use ScratchABlock?

### ISSUE THREAD

**Issue by wgaylord** [pfalcon/ScratchABlock#6]
I can't find how I am to actually run this. Also what else do I need to install. (Considering I know you had to had use other lib then the standard built-ins)

(BTW I am attempting to run this on Windows)

**Comment by pfalcon** [comment]
Did you try to read README?

**Comment by wgaylord** [comment]
Yes. No where does it actually say how to use it. Only which file is the main file. (Which when you try to run it under python 3 it throws import errors)

I have fixed the one import error but a new error has emerged.  

> C:\Users\chibi\Desktop\Decompiler\ScratchABlock>python3 apply_xform.py
> Traceback (most recent call last):
>   File "apply_xform.py", line 18, in <module>
>     from asmprinter import AsmPrinter
>   File "C:\Users\chibi\Desktop\Decompiler\ScratchABlock\asmprinter.py", line 9, in <module>
>     class AsmPrinter(CFGPrinter):
> NameError: name 'CFGPrinter' is not defined

Information about my system:
Windows 10 x64
Python 3.6.0
Manually install pyyaml as you require it but don't list it anywhere.


Also why is it that devs are so nasty towards people asking for help?

Also from reading what ScratchABit says I see you offer no support normally for windows..... Seems much like you uphold what you are against.

**Comment by mewmew** [comment]
@chibill I have limited experience with how dependencies are satisfied in Python, but I tried running `apply_xform.py` on my system, using Python 3.6.1 and as you pointed out, first I got an error that it was unable to locate the `yaml` import. After installing [pyyaml](http://pyyaml.org/) it works for me.

I'm running Arch Linux with Python 3.

```
u@x220 ~/D/ScratchABlock> python3 apply_xform.py 
usage: apply_xform.py [-h] [-o OUTPUT] [--script SCRIPT] [--iter]
                      [--funcdb FUNCDB] [--format {none,bblocks,asm,c}]
                      [--no-dead] [--no-graph-header] [--annotate-calls]
                      [--repr] [--debug] [--log-level LOG_LEVEL]
                      file
apply_xform.py: error: the following arguments are required: file
```

So it seems like the Python setup you are using is not working for whatever reason. CFGPrinter is defined in core.py and apply_xform.py imports everything from core with `from core import *`

Sorry to not be able to give any specific leads on what may be wrong, but just wanted to let you know that I could get it running simply by installing pyyaml on Arch Linux, and so it should be just something with how python handles local import paths on your system or something along those lines.

Best of luck and happy reversing.

Cheers /u

**Comment by wgaylord** [comment]
Found the problem 
https://github.com/pfalcon/ScratchABlock/blob/master/asmprinter.py#L4

he doesn't do from core import *
he has import core

I have realized I can't use this project at all because the other half of it only runs on Linux.

**Comment by mewmew** [comment]
> Found the problem
> https://github.com/pfalcon/ScratchABlock/blob/master/asmprinter.py#L4
> 
> he doesn't do from core import *
> he has import core

I see, good that you found the issue.

> I have realized I can't use this project at all because the other half of it only runs on Linux.

I see. Personally, I would be curious to learn what parts of ScratchABlock are platform independent, and what parts are relying on Linux (POSIX/Unix) interfaces.

@pfalcon Care to elaborate?

**Comment by mewmew** [comment]
> I have realized I can't use this project at all because the other half of it only runs on Linux.

@chibill Also, may you wish to be more specific, which other half of it?

**Comment by wgaylord** [comment]
SrcatchABit

Which you need to be able to make any use of ScratchABlock.
Unless you think I should write a disassembler myself. Which is a bit goofy.

**Comment by mewmew** [comment]
Ah, I see. So it's ScratchABit that is Linux/Unix specific. Thanks for clearing that up.

**Comment by pfalcon** [comment]
> Manually install pyyaml as you require it but don't list it anywhere.

Thanks for the report, fixed in f1e9f52ec11a0935159125994984c34427176e49.

> Also why is it that devs are so nasty towards people asking for help?

Can you elaborate what you mean?

> Yes. No where does it actually say how to use it. 

Hmm, let's read it together:

> tests - The main testsuite. While integrational in the nature, it usually tests one pass on one simple file, so follows the unit testing philosophy. Tests are represented as PseudoC input files, while expected results - as PseudoC with basic blocks annotation and (where applicable) CFG in .dot format. **Looking at these testcases, trying to modify them and seeing the outcome is the best way to learn how ScratchABlock works.**

So, the README explains how to use ScratchABlock, you read it, you ask question which doesn't makes much sense if you read the README, you being asked if you read it, instead of reading it, you start to call names. Funny.

**Comment by pfalcon** [comment]
> NameError: name 'CFGPrinter' is not defined

I can't reproduce this issue:

~~~
pfalcon@x230:~/tmp/ScratchABlock$ uname -a
Linux x230 4.8.0-56-generic #61~16.04.1-Ubuntu SMP Wed Jun 14 11:58:22 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
pfalcon@x230:~/tmp/ScratchABlock$ python3 --version
Python 3.5.2
pfalcon@x230:~/tmp/ScratchABlock$ ./apply_xform.py
usage: apply_xform.py [-h] [-o OUTPUT] [--script SCRIPT] [--iter]
                      [--funcdb FUNCDB] [--format {none,bblocks,asm,c}]
                      [--no-dead] [--no-graph-header] [--annotate-calls]
                      [--repr] [--debug] [--log-level LOG_LEVEL]
                      file
apply_xform.py: error: the following arguments are required: file
~~~

**Comment by pfalcon** [comment]
> I see. Personally, I would be curious to learn what parts of ScratchABlock are platform independent, and what parts are relying on Linux (POSIX/Unix) interfaces.
> @pfalcon Care to elaborate?

Everything is fully portable and doesn't use any OS-specific features (unless proven otherwise with a reproducible test report).

**Comment by pfalcon** [comment]
> SrcatchABit
> Which you need to be able to make any use of ScratchABlock.

You don't need ScratchABit to use ScratchABlock. You *can*, but you don't have to.

> Unless you think I should write a disassembler myself. Which is a bit goofy.

Of course you should. Unless you think that *my* disassembler will be of use to *you*. Which is yet another silent assumption you appear to make. What are you trying to achieve? README pretty clearly states that ScratchABlock is experimental, research project, it won't work like you want, you'll need to learn it and make it work like you want. ScratchABit is more "ready", but it still does what I need, to make it do what you need, you'll need to do legwork just the same.

**Comment by wgaylord** [comment]
So basically you don't think I should you your work.... 

If you didn't want people using it why the heck did you put it on github....

**Comment by pfalcon** [comment]
Again some random assumptions. I spent on this projects 2 years, and want you to spend on it at least 2 years, fixing all the issues you find and developing it further. That's what I want. But nobody cares what somebody else wants, and everyone is free to do what they want with what they have on their hands.

**Comment by wgaylord** [comment]
You one assuming stuff is you. You literally said go make your own. Which by logic means you don't want me using yours

**Comment by wgaylord** [comment]
Messed are between you and one

**Comment by pfalcon** [comment]
These projects are at most 25% (maybe 10%) of what I'd like them to be, and they don't even satisfy all of my own needs. You'll need to do your own legwork to see if they're useful for you at all in their current state, and then 75% of missing things still will need to be done, and you would need to be prepared to that. Chances, you won't find them much useful to you. At least, I looked at ~20 of other projects at found them useless, that's why I started with ScratchABit/ScratchABlock myself. Surely, I may anticipate that someone else may find mine not useful just the same. I share them in the hope some still might find them useful, but as is and without any warranty (I'll try to fix any bugs properly reported though).
